### PR TITLE
Fix and add additional tests to test suite

### PR DESCRIPTION
**Pull request by sethrj**
There are a few "include what you use" errors on newer test cases that didn't get added in #1398, and I have also added some test cases that are used to help test the Fortran language but are not Fortran-specific. For example, the `director_simple` test would be helpful for packages trying to dip their feet into supporting director methods, since most of the existing tests are quite complex and meant to test edge cases rather than basic functionality.

### CODE REVIEW

File: Examples/test-suite/varargs.i [hunk 1]
```diff
@@ -8,6 +8,10 @@
 #include <string.h>
 %}
 
+%{
+#include <string.h>
+%}
+
```
Comment (ojwb) [inline Examples/test-suite/varargs.i:14]: This duplicates what's already present just above.

File: Lib/swigfragments.swg [hunk 2]
```diff
@@ -88,3 +88,11 @@
...
 #include <memory>
 %}
+
+%fragment("<utility>", "header") %{
+#include <utility>
+%}
+
+%fragment("<cctype>", "header") %{
+#include <cctype>
+%}
```
Comment (ojwb) [inline Lib/swigfragments.swg:98]: This new `<cctype>` fragment doesn't actually seem to be used (nor does `<cctype>` anywhere in the SWIG source in fact).

File: Examples/test-suite/javascript/Makefile.in [hunk 3]
```diff
@@ -50,6 +50,7 @@ ifeq (node,$(JSENGINE))
   # contract macro has an empty 'else' at the end...
   aggregate.cpptest: GYP_CFLAGS = \"-Wno-empty-body\"
   contract.cpptest: GYP_CFLAGS = \"-Wno-empty-body\"
+  contract_c.ctest: GYP_CFLAGS = \"-Wno-empty-body\"
```
[the remark this answers is not in the document]
Comment (ojwb) [reply Examples/test-suite/javascript/Makefile.in:53]: PR for that is #2206

[the remark this answers is not in the document]
Comment (ojwb) [reply Examples/test-suite/javascript/Makefile.in:53]: The CI passed so I've merged that PR now.

**Comment by sethrj** [comment]
@ojwb Thanks for the feedback and warning fixes. I've addressed all your comments and merged master into this branch. Please let me know if you'd prefer that I rebase. (Based on previous reviews it sounds like merge is preferred in the SWIG review workflow, and rebase can tend to "eat" review comments in Git.)

**Comment by ojwb** [comment]
> Please let me know if you'd prefer that I rebase.

Sounds good - I think for preserving the branch history isn't useful here, but github says it can't rebase and I suspect it's not a clean rebase at the command line either, so better you do it as you're more familiar with the change.

> (Based on previous reviews it sounds like merge is preferred in the SWIG review workflow, and rebase can tend to "eat" review comments in Git.)

I think it's really "in Github" rather than "in Git".

I'm not sure we have a firm project position on it, but I'm not a big fan of rebasing during review.  Sometimes it's the best approach though (e.g. when reviving a long-lived PR or when part of the change gets merged first or when the changes in the patch are very small).

For the final merge I vaguely recall @wsfulton has (or at least had in the early days of SWIG using git) a preference to keep master history simple.  In recent times I've used whichever of squash, rebash or merge seemed most appropriate for a particular branch and not had any complaints.  My experience with git in general has been that it's really the long lived branches where the branch off and merge back crosses many other branch merges and hundreds or thousands of commits which are actually hard to get your head around when digging through history later, and that it's sometimes useful to be able to see the history of development within the branch when trying to work out why something was done as it was.

**Comment by wsfulton** [comment]
If you read most of the commit messages in this pull request there is no context or explanation what the changes are about nor are there any code commits associated with the changes. This is quite unusual, so a reason why the test is being modified needs to be clearly stated in the commit message.

Generally rebasing is a bad idea if a branch is being monitored/reviewed, but not always. I seem to recall a rebase on the Fortran branch completely wrecking monitoring the progress and review up to that date. However, in this case the branch is tiny and most of the commits on this branch don't make much sense so I suggest you rebase it to fix the commit messages so that we can review each commit from fresh.

**Comment by sethrj** [comment]
> I think it's really "in Github" rather than "in Git".

Quite so, I meant to say GitHub.

> For the final merge I vaguely recall @wsfulton has (or at least had in the early days of SWIG using git) a preference to keep master history simple. In recent times I've used whichever of squash, rebash or merge seemed most appropriate for a particular branch and not had any complaints. My experience with git in general has been that it's really the long lived branches where the branch off and merge back crosses many other branch merges and hundreds or thousands of commits which are actually hard to get your head around when digging through history later, and that it's sometimes useful to be able to see the history of development within the branch when trying to work out why something was done as it was.

I'm more or less in agreement. The nice thing about linear history is it makes bisecting and `git log` far easier since each single commit is meaningful and tested. With GitHub's "squash and merge" feature it allows the branch development to remain continuous and not have to worry about cleaning up minor changes, fixups, merges from master, etc.

**Comment by sethrj** [comment]
@wsfulton Thanks for the feedback, I've reworked this entire branch, deleting a couple of changes that I realize are no longer useful for the Fortran branch (viz. the `<utility>` fragment), adding python test cases for a few tests, and merging a couple of the new tests into existing tests. I hope you find the new commits easier to read and sufficiently documented.

**Comment by sethrj** [comment]
Note that previously I'd disabled the "multi-import" test extension in Go, but didn't mark why. [This test failure](https://github.com/swig/swig/runs/5170606404) shows that Go seems to have a problem with the test:
```
# swigtests/multi_import_b
[1260](https://github.com/swig/swig/runs/5170606404?check_suite_focus=true#step:7:1260)
./multi_import_b.go:47:8: imported and not used: "swigtests/multi_import_d"
[1261](https://github.com/swig/swig/runs/5170606404?check_suite_focus=true#step:7:1261)
# swigtests/multi_import_b
[1262](https://github.com/swig/swig/runs/5170606404?check_suite_focus=true#step:7:1262)
../multi_import_b/multi_import_b.go:47:8: imported and not used: "swigtests/multi_import_d"
[1263](https://github.com/swig/swig/runs/5170606404?check_suite_focus=true#step:7:1263)
# swigtests/multi_import_b
[1264](https://github.com/swig/swig/runs/5170606404?check_suite_focus=true#step:7:1264)
../multi_import_b/multi_import_b.go:47:8: imported and not used: "swigtests/multi_import_d"
[1265](https://github.com/swig/swig/runs/5170606404?check_suite_focus=true#step:7:1265)
make[1]: *** [Makefile:69: multi_import.multicpptest] Error 2
```

I'm extending the test to try to fix this error.

**Comment by ojwb** [comment]
See https://github.com/swig/swig/issues/1599#issuecomment-538599753 - the mechanism described there would presumably need to resolve which parent class a method is inherited from in complex multiple inheritance situations, so maybe it just doesn't always get that right currently.

**Comment by ojwb** [comment]
Oh, actually the only CI failing now is appveyor, so looks like you got the test to pass with Go.

**Comment by ojwb** [comment]
I know - what I meant was I'd seen the CI was red and had assumed you were still fighting the Go issue, hence the first comment, but then I spotted it was just the usual appveyor brokenness.

**Comment by sethrj** [comment]
Gotcha. Yes, the issue with Go is that it fails if you import a module but don't use any of it, and I had an absurd time figuring out how to get Go to use the module in the diamond while getting all the target languages compiling without linking against any libraries...  anyway, it's figured out now.

**Comment by ojwb** [comment]
> the issue with Go is that it fails if you import a module but don't use any of it

Ah yes, I even knew that but hadn't made the connection that this might be an issue in SWIG testcases.

> Once this is merged, the pending Fortran work will have _only_ changes related to the core Fortran module, tests, and documentation.

Progress!

> Would you like me to resurrect #1195 , or create a fresh pull request?

I'd lean towards keeping it in the same place at least in this case.

**Comment by sethrj** [comment]
Thanks @ojwb!

**Comment by sethrj** [comment]
I added that because of a bug in the Fortran module early in its development: putting the "ignore" conditional in the wrong place meant that the typemap was being ignored.

I'm sorry I didn't see the warnings: I will patch them immediately.

**Comment by ojwb** [comment]
Oh sorry, I didn't see the warning either.  It's really too easy to miss such new warnings - to thoroughly check the CI output for a proposed patch would mean having to click on 100+ job entries to check the log for each.  Even just checking one job for each supported language would be a lot of clicks, and easy to miss one.

@wsfulton I suggest we add `-Werror` to the SWIG command line for the testsuite (except for `make check-errors-test-suite` obviously), and make "no unsuppressed warnings from the testsuite" a hard requirement rather than just something we try for.

**Comment by wsfulton** [comment]
No big deal. The test-suite didn't fail after the commit (which is the important thing) and I'm quite used to mopping up warnings for various things. Thanks for sorting out promptly.

Regarding using `swig -Werror`, that would be nice, but it's not so simple. There is a practical problem in that even for the standard (non-experimental) languages, it is quite acceptable to have missing features. Warnings often appear due to missing features. We use `%warnfilter` in the test-suite as the correct way to clean up the test-suite output for valid warnings. I wouldn't advocate using %warnfilter to suppress warnings that need addressing. So, allowing SWIG warnings does actually help as it shows where problems need addressing. This approach allows language modules of varying maturity and incremental development of a language module from experimental status to supported status through to being a mature, best-in-class module. 

Open to suggestions to address though. Maybe, if all the current warnings can be fixed for the non-experimental languages, given there aren't actually that many, then we could add a warning free run as an extra requirement for obtaining supported status and at the expense of less flexibility.
